### PR TITLE
Mark Bugsnag frames as out of project

### PR DESCRIPTION
**Pull request by Cawllec**
## Goal
To enable easier examination of the Bugsnag reporting process, and why certain events appear as they do. 

## Design

Rather than dropping frames from `lib/bugsnag` they are now included, but will automatically be marked out of project.

## Tests

Unit test added to ensure Bugsnag frames are added correctly and `inProject` is not set.

Will require #496 to be merged first so that Maze tests can be added (in progress)

## Review

For the submitter, initial self-review:

- [ ] Commented on code changes inline explain the reasoning behind the approach
- [ ] Reviewed the test cases added for completeness and possible points for discussion
- [ ] A changelog entry was added for the goal of this pull request
- [ ] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [ ] Initial review of the intended approach, not yet feature complete
  - [ ] Structural review of the classes, functions, and properties modified
  - [ ] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: spec/report_spec.rb [hunk 1]
```diff
@@ -512,14 +512,15 @@ def gloops
     }
   end
 
-  it "marks the top-most stacktrace line as inProject if necessary" do
+  it "marks the top-most non-bugsnag stacktrace line as inProject if necessary" do
```
Comment (snmaynard) [inline spec/report_spec.rb]: This shouldnt be required

File: spec/report_spec.rb [hunk 2]
```diff
@@ -512,14 +512,15 @@ def gloops
...
     Bugsnag.configuration.project_root = File.expand_path File.dirname(__FILE__)
     Bugsnag.notify(BugsnagTestException.new("It crashed"))
```
Comment (snmaynard) [inline spec/report_spec.rb]: We should raise this exception so it has a populated stacktrace

File: spec/report_spec.rb [hunk 3]
```diff
@@ -512,14 +512,15 @@ def gloops
...
 
     expect(Bugsnag).to have_sent_notification{ |payload, headers|
       exception = get_exception_from_payload(payload)
       expect(exception["stacktrace"].size).to be >= 1
-      expect(exception["stacktrace"].first["inProject"]).to eq(true)
+      top_frame = get_project_frame(exception["stacktrace"])
```
Comment (snmaynard) [inline spec/report_spec.rb]: This should then be able to operate on the stacktrace directly

File: spec/report_spec.rb [hunk 4]
```diff
@@ -1051,11 +1052,11 @@ def gloops
       expect(exception["errorClass"]).to eq("RuntimeError")
       expect(exception["message"]).to eq("'nil' was notified as an exception")
 
-      stacktrace = exception["stacktrace"][0]
-      expect(stacktrace["lineNumber"]).to eq(1047)
+      stacktrace = get_project_frame(exception["stacktrace"])
```
Comment (snmaynard) [inline spec/report_spec.rb]: This should be moved into a separate test - as this test is just testing the message. We should probably test that the stacktrace is bugsnag initially then this file after the right number of frames. But that might be covered by what is in the stacktrace spec now?

File: spec/stacktrace_spec.rb [hunk 5]
```diff
@@ -25,6 +26,21 @@
     }
   end
 
+  it "includes bugsnag lines marked out of project" do
```
Comment (snmaynard) [inline spec/stacktrace_spec.rb]: I dont think this should be here. I think its a test of report and not stacktrace.

File: spec/stacktrace_spec.rb [hunk 6]
```diff
@@ -12,8 +12,9 @@
 
     expect(Bugsnag).to have_sent_notification{ |payload, headers|
       exception = get_exception_from_payload(payload)
-      starting_line = __LINE__ - 10
-      expect(exception["stacktrace"][1]["code"]).to eq({
+      starting_frame = get_project_frame(exception["stacktrace"], 1)
```
Comment (snmaynard) [inline spec/stacktrace_spec.rb]: Instead of using this function, use a function that explicitly raises and catches an exception to notify - rather than notifying an exception with no stack.

File: spec/spec_helper.rb [hunk 7]
```diff
@@ -84,3 +84,18 @@ def have_sent_notification(&matcher)
     end
   end
 end
+
+def get_project_frame(stacktrace, count=0)
```
Comment (snmaynard) [inline spec/spec_helper.rb]: I dont think this should be required

File: lib/bugsnag/stacktrace.rb [hunk 8]
```diff
@@ -27,7 +27,7 @@ def initialize(backtrace, configuration)
         # Parse the stacktrace line
 
         # Skip stacktrace lines inside lib/bugsnag
```
Comment (tobyhs) [inline lib/bugsnag/stacktrace.rb]: I think this comment is no longer applicable

**Comment by Cawllec** [comment]
I've fixed the maze tests that this change has broken.

File: features/plain_features/report_stack_frames.feature [hunk 9]
```diff
@@ -14,8 +14,8 @@ Scenario Outline: Stack frames can be removed
   And the request used the "Ruby Bugsnag Notifier" notifier
   And the request used payload v4 headers
   And the request contained the api key "a35a2a72bd230ac0aa0f52715bbdc6aa"
-  And the "file" of stack frame 0 equals "/usr/src/app/stack_frame_modification/initiators/<initiator>.rb"
```
Comment (snmaynard) [inline features/plain_features/report_stack_frames.feature]: You should instead update the features to raise the exception so that bugsnag doesnt appear in the stacktrace

**Comment by Cawllec** [comment]
Additional maze changes have been made to ensure stacks from different types of handled errors (a string vs a caught exception) are both tested.

File: features/steps/ruby_notifier_steps.rb [hunk 10]
```diff
@@ -12,4 +12,30 @@
   steps %Q{
     When I set environment variable "#{env_var}" to "#{credentials}@#{current_ip}:#{MOCK_API_PORT}"
   }
+end
+Then(/^the "(.+)" of the top project stackframe equals "(.+)"(?: for request (\d+))?$/) do |element, value, request_index|
+  stacktrace = read_key_path(find_request(request_index)[:body], 'events.0.exceptions.0.stacktrace')
+  frame_index = 0
```
Comment (tobyhs) [inline features/steps/ruby_notifier_steps.rb]: I think you can use `Array#find_index` to make this more concise (probably something like `frame_index = stacktrace.find_index { |frame| frame['file'] !~ %r{lib/bugsnag.*\.rb\z} }`)

File: features/steps/ruby_notifier_steps.rb [hunk 11]
```diff
@@ -12,4 +12,30 @@
   steps %Q{
     When I set environment variable "#{env_var}" to "#{credentials}@#{current_ip}:#{MOCK_API_PORT}"
   }
+end
+Then(/^the "(.+)" of the top project stackframe equals "(.+)"(?: for request (\d+))?$/) do |element, value, request_index|
```
Comment (tobyhs) [inline features/steps/ruby_notifier_steps.rb]: Can we have a blank line between step definitions for organization/easier editor navigation? (between lines 15-16, 28-29)

File: features/steps/ruby_notifier_steps.rb [hunk 12]
```diff
@@ -12,4 +12,30 @@
...
+  stacktrace = read_key_path(find_request(request_index)[:body], 'events.0.exceptions.0.stacktrace')
+  frame_index = 0
+  stacktrace.each do |frame, index|
+    unless /.*lib\/bugsnag.*\.rb/.match(frame["file"])
```
Comment (tobyhs) [inline features/steps/ruby_notifier_steps.rb]: The step definition pattern has "of the top project stackframe", so should this actually check the `inProject` value instead?

Otherwise, if checking against this regular expression is correct, should the step definition pattern have something like "of the top non-Bugsnag stackframe" instead?

File: features/steps/ruby_notifier_steps.rb [hunk 13]
```diff
@@ -12,4 +12,30 @@
...
+      frame_index = index
+      break
+    end
+  end
+  steps %Q{
+    the "#{element}" of stack frame #{frame_index} equals "#{value}"
+  }
+end
+Then(/^the "(.+)" of the top project stackframe equals (\d+)(?: for request (\d+))?$/) do |element, value, request_index|
```
Comment (tobyhs) [inline features/steps/ruby_notifier_steps.rb]: I don't think you need 2 very similar step definitions; you can probably get away with using `(\d+|(".+"))` (so the double quotes are part of the capture for the string case and you can just have `equals #{value}`)

File: features/plain_features/report_stack_frames.feature [hunk 14]
```diff
@@ -14,33 +14,49 @@ Scenario Outline: Stack frames can be removed
...
+  And I wait for 1 second
+  Then I should receive a request
+  And the request used the "Ruby Bugsnag Notifier" notifier
+  And the request used payload v4 headers
+  And the request contained the api key "a35a2a72bd230ac0aa0f52715bbdc6aa"
+  And the "file" of the top non-bugsnag stackframe equals "/usr/src/app/stack_frame_modification/initiators/handled_block.rb"
+  And the "lineNumber" of the top non-bugsnag stackframe equals <lineNumber>
+
+  Examples:
+  | ruby version | lineNumber |
```
[the remark this answers is not in the document]
Comment (Cawllec) [reply features/plain_features/report_stack_frames.feature]: Removed

File: features/plain_features/handled_errors.feature [hunk 15]
```diff
@@ -67,8 +67,8 @@ Scenario Outline: A handled error can attach metadata in a block
   And the event "severity" equals "warning"
   And the event "severityReason.type" equals "handledException"
   And the exception "errorClass" equals "RuntimeError"
-  And the "file" of stack frame 0 equals "/usr/src/app/handled/block_metadata.rb"
```
[the remark this answers is not in the document]
Comment (Cawllec) [reply features/plain_features/handled_errors.feature]: Removed, and broken the top most scenario out into two separate scenarios for the same reason.

File: features/plain_features/report_stack_frames.feature (hunk 9 above)
Comment (Cawllec) [reply features/plain_features/report_stack_frames.feature]: I've updated the scenarios to separate the tests that require this change and those that don't.
